### PR TITLE
JS: Refactor Promises.qll to use PreCallGraphStep

### PR DESCRIPTION
**Pull request by erik-krogh**

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/Promises.qll [hunk 1]
```diff
@@ -283,188 +255,143 @@ private module PromiseFlow {
...
-    override predicate loadStore(DataFlow::Node pred, DataFlow::Node succ, string prop) {
+      pred = call.getCallback([0 .. 1]).getAReturn()
+    )
+    or
+    // promise.catch()
+    exists(DataFlow::MethodCallNode call | call.getMethodName() = "catch" and succ = call |
       prop = valueProp() and
-      pred = getReceiver().getALocalSource() and
-      succ = this
+      pred = call.getReceiver().getALocalSource()
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/Promises.qll:379]: I understand this code was already like that, but why use `getALocalSource` here? It seems like it should just be `call.getReceiver`.

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/Promises.qll:379]: If we don't include `.getALocalSource()`, we will miss e.g. [this sink](https://github.com/github/codeql/blob/master/javascript/ql/test/library-tests/Promises/flow.js#L84).  (I ran the test)

I actually don't remember why it is needed.  
But I think its related that `succ` in `storeStep` must be a `SourceNode`.

Comment (asgerf) [reply javascript/ql/src/semmle/javascript/Promises.qll:379]: That's really creepy. The path to that sink shouldn't even need this step all (this step's successor is the `catch` call itself, which in that test is a dead end). You're sure it's not just the `nodes/edges` relations that changed?

In any case, it's not a fault introduced by this PR. Happy to move the investigation into an issue if you prefer.

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/Promises.qll]: > You're sure it's not just the nodes/edges relations that changed?

I thought I had checked that, but apparently I hadn't.  
It was just an `edges` relation that changed (actually a type-tracking-step relation).

**Comment by erik-krogh** [comment]
> Hope the evaluation works out.

It did: https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1593134229710

**Pull request merged**